Maschine mk2 mikro

nogizhopaboroda opened the pull request:
- Add Maschine MK2 Mikro mappings
  - Add demos for MK2 Mikro
  - Extend rgb_led component
  - Add OLED Display component
  - Add pads component

ni-azhukov committed:
feat(maschine mk2 mikro):
  - Add Maschine MK2 Mikro mappings
  - Add demos for MK2 Mikro
  - Extend rgb_led component
  - Add OLED Display component
  - Add pads component

asutherland commented:
Thank you for the pull request!

Can you clarify what the contents of `monroe.json` are?  If it's an image converted directly to JSON data, we do already have the jpeg dependency for the Mk3 examples that could be used and allow people to try out their own custom image displays more easily.  And if we are using an image/converted image, it'd be great to make sure the provenance of the image is documented, like is done in https://github.com/asutherland/ni-controllers-lib/commit/f698d54bb1f33c0a8b7b55a92c13da8ceff3ce31 for the mk3 demo image.  Thanks!

ni-azhukov committed:
feat(oled display):
  - Add drawPixel method
  - Add drawBitmap method
  - Add drawImage method
  - Load placeholder image from file

nogizhopaboroda commented:
Hi @asutherland 

Thanks for your comment. I updated the pr (commit https://github.com/nogizhopaboroda/ni-controllers-lib/commit/d7b1e96457235c0a2249f2c3dbb1e6dcd9fd891d). Now i'm rendering placeholder jpeg file as it is in mk3 demo.
Please check it out if you have a chance.

asutherland commented on lib/maschine_mk2_mikro_config.json:
```diff
@@ -0,0 +1,713 @@
+  "productId": 4608,
+  "name": "Maschine mk2 mikro",
+
+  "ledBrightestValue": 63,
```
Can you confirm that 63 is right for this instead of 127?  For my other NI hardware with RGB LED's the right value was 127 so this might leave the LEDs only ever at half brightness?  I have a (non-mikro) Maschine Mk2 in a drawer somewhere I can get out and take a look if it's likely they work the same.

nogizhopaboroda replied:
right, thanks! they were really at half brightness!

asutherland reviewed (changes requested):
Thank you for updating the example image per my request!

And thanks again for the pull request and apologies about the delay in getting back to this.  This is basically good to land modulo a few minor changes.  (And I can make those changes if need be.)

ni-azhukov committed:
fix(led rgb): change setColorHelper signature, accept brightness argument

ni-azhukov committed:
fix(mk3 demo): remove leftovers from demo

ni-azhukov committed:
fix(basde controller): fix typo

nogizhopaboroda commented on lib/components/input/pads.ts:
```diff
@@ -0,0 +1,64 @@
+  prefix: string;
+}
+
+export class Pads implements Widget {
```
the implementation here is taken from https://github.com/shaduzlabs/cabl/blob/master/src/devices/ni/MaschineMikroMK2.cpp#L332-L357
where 
`kMikroMK2_padDataSize = 64`
https://github.com/shaduzlabs/cabl/blob/master/src/devices/ni/MaschineMikroMK2.h
also confusing to me

asutherland replied:
Thanks for the clarification.  This shall remain a mystery for now! ;)

ni-azhukov committed:
fix(mk2 config): merge outputs + increase led brightest value

nogizhopaboroda commented:
Hi @asutherland ,

Thanks for your input!
i addressed your comments, please check again.

nogizhopaboroda commented on lib/components/output/led_rgb.ts:
```diff
@@ -38,4 +57,21 @@ export class LED_RGB {
+    const color = tinycolor({ r: this.r, g: this.g, b: this.b });
+    const { r, g, b } = color.brighten(f).toRgb();
+    this.setRGB(r, g, b);
+  }
```
added `brightness` argument. a bit confusing the difference of the interfaces between `led_indexed` and `led_rgb` now, but i agree it's better this way

asutherland reviewed (approved):
Woo!  I'll land this now and then push some npm buttons to make a new release.

asutherland commented:
I published this in npm as version 1.1.0 after making the changes included in https://github.com/asutherland/ni-controllers-lib/commit/9b0064868ed7e58603151b5234b7b509a1847c94 where the main thing was that the oled displays firstByte config was a hex string but was being directly used as a number, so I converted that to the raw numeric value.  Hopefully things still work the same or work better!
